research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -0,0 +1,368 @@
+# Lint as: python3
+# Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Training utilities for Global Features model."""
+
+import os
+import pickle
+import time
+
+import numpy as np
+import tensorflow as tf
+
+from delf.python import whiten
+from delf.python.datasets.revisited_op import dataset as test_dataset
+from delf.python.datasets.sfm120k import sfm120k
+from delf.python.training import global_features_utils
+from delf.python.training.model import global_model
+
+
+def grad(criterion, model, input, target, neg_num=5):
+  """Records gradients and loss through the network.
+
+  Args:
+    criterion: Loss function.
+    model: Network for the gradient computation.
+    input: Tuple of query, positive and negative images.
+    target: List of indexes to specify queries (-1), positives(1), negatives(0).
+    neg_num: Integer, number of negatives per a tuple.
+
+  Returns:
+    loss: Loss for the training step.
+    gradients: Computed gradients for the network trainable variables.
+  """
+  # Record gradients and loss through the network.
+  with tf.GradientTape() as tape:
+    output = tf.Variable(
```

Comment:
maybe rename `output` to `descriptors` would make it easier to understand. Here in particular, since the fn is computing the gradients, I was expecting output to be gradients.